### PR TITLE
license plate detector

### PR DESCRIPTION
**Pull request by ongzizhao**
A new peekingduck node for license plate detection, single class object detector
Custom trained Yolov4/v4tiny model using pretrained weights from Alexey Repo.

Files created:
configs/model/license_plate_detector.yml
configs/draw/blur_bbox.yml
model.license_plate_detector.py
model.licenseplate.lp_detector.model.py
model.licenseplate.licenseplate_files.detector.py

Unit test also created for the two nodes
test_LPdetector.py
test_blur_bbox.py

### CODE REVIEW

File: peekingduck/configs/model/license_plate_detector.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    v4 : ../weights/yolo_license_plate/yolov4-LP.cfg
+}
+model_weights: {
+    v4tiny: ../weights/yolo_license_plate/yolov4tiny-LP.weights,
+    v4: ../weights/yolo_license_plate/yolov4-LP.weights
+}
+
+size: 416
+confThreshold: 0.1
+nmsThreshold: 0.3
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/configs/model/license_plate_detector.yml:17]: ok i will change the config variable name to align with the existing Yolo node

File: peekingduck/pipeline/nodes/model/license_plate_detector.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+License Plate Detection model
+"""
+
+from typing import Dict, Any
+from peekingduck.pipeline.nodes.node import AbstractNode
+from .licenseplate import lp_detector_model
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/license_plate_detector.py:21]: update with absolute import

File: peekingduck/pipeline/nodes/model/license_plate_detector.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            dictionary pointing to path of the model configuration file.
+
+        model_weights (:obj:`Dict`):
+            dictionary pointing to path of the model weights file.
+
+        size (:obj:`int`): **default = 416 **
+
+            image resolution passed to the YOLO model.
+
+        confThreshold (:obj:`float`): **[0,1], default = 0.1**
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/license_plate_detector.py:63]: The variable names updated to align with existing YOLO

File: peekingduck/pipeline/nodes/model/license_plate_detector.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+            bounding box with confidence score less than the specified
+            confidence score threshold is discarded.
+
+        nmsThreshold (:obj:`float`): **[0,1], default = 0.3**
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/license_plate_detector.py:68]: The variable names updated to align with existing YOLO

File: peekingduck/pipeline/nodes/model/license_plate_detector.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    def __init__(self, config: Dict[str, Any] = None, **kwargs: Any) -> None:
+        super().__init__(config, node_path=__name__, **kwargs)
+        self.model = lp_detector_model.Yolov4(self.config)
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        """function that reads the image input and returns the bboxes
+        of the specified objects chosen to be detected
+        Args:
+            inputs (Dict): Dictionary of inputs with key "img"
+        Returns:
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/license_plate_detector.py:94]: done

File: peekingduck/pipeline/nodes/model/licenseplate/licenseplate_files/detector.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+        args:
+            - yolo:  (Model) model like yolov4 or yolov4_tiny
+            - image: (np.array) input image
+
+        return:
+            - boxes: (np.array) an array of bounding box with
+                    definition like (x1, y1, x2, y2), in a
+                    coordinate system with original point in
+                    the left top corner
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/licenseplate/licenseplate_files/detector.py:104]: made the changes

File: peekingduck/pipeline/nodes/draw/blur_bbox.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import numpy as np
+from scipy.ndimage import gaussian_filter
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+
+class Node(AbstractNode):  # pylint: disable=too-few-public-methods
+    """Blur area bounded by bounding boxes on image.
+
+    The draw blur_bbox node uses the bboxes and blur the area of the image
+    bounded by the bboxes.
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py:28]: edited the doc string

File: peekingduck/pipeline/nodes/draw/blur_bbox.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    Inputs:
+
+        |img|
+
+        |bboxes|
+
+    Outputs:
+        |none|
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py:37]: the output is an img, made the appropriate changes

File: peekingduck/pipeline/nodes/draw/blur_bbox.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            image[y_1:y_2, x_1:x_2, :] = blur_bbox_image
+
+        return image
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        """function that reads the image input and returns the bboxes
+        of the specified objects chosen to be detected
+        Args:
+            inputs (Dict): Dictionary of inputs with keys "img", "bboxes"
+        Returns:
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py:72]: done

File: peekingduck/pipeline/nodes/draw/blur_bbox.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            # apply the blur using gaussian filter from scipy
+            blur_bbox_image = gaussian_filter(bbox_image, sigma=5)
+            image[y_1:y_2, x_1:x_2, :] = blur_bbox_image
+
+        return image
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        """function that reads the image input and returns the bboxes
+        of the specified objects chosen to be detected
+        Args:
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py:70]: done

File: peekingduck/pipeline/nodes/model/license_plate_detector.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+from typing import Dict, Any
+from peekingduck.pipeline.nodes.node import AbstractNode
+from .licenseplate import lp_detector_model
+
+
+class Node(AbstractNode):  # pylint: disable=too-few-public-methods
+    """Yolo node class that initialises and use yolo model to infer bboxes
+    from image frame.
+
+    The yolo node is capable of detecting objects from 80 categories. It uses
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/license_plate_detector.py:28]: doc string updated to describe single class(License Plate) detector

File: peekingduck/pipeline/nodes/model/license_plate_detector.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    YOLOv4-tiny by default and can be changed to using YOLOv4. The table of
+    categories can be found :term:`here <object detection indices>`.
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/license_plate_detector.py:30]: The portion on table of categories is removed from doc string

File: peekingduck/pipeline/nodes/model/licenseplate/licenseplate_files/detector.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        )
+
+        return model
+
+    @staticmethod
+    def bbox_scaling(bboxes: List[list], scale_factor: float) -> List[list]:
+        """
+        To scale the width and height of bboxes from v4tiny
+        After conversion of model from .cfg .weights from ,Alexey's Darknet repo,
+        to tf model, bboxes are bigger. So downscaling required for better fit
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/model/licenseplate/licenseplate_files/detector.py:77]: made changes to doc string

File: tests/conftest.py [hunk 14]
```diff
@@ -23,6 +23,8 @@
 
 TEST_HUMAN_IMAGES = ['t1.jpg', 't2.jpg', 't4.jpg']
 TEST_NO_HUMAN_IMAGES = ['black.jpg', 't3.jpg']
+TEST_NO_LP_IMAGES = ['black.jpg', 't3.jpg']
+TEST_LP_IMAGES = ['tcar1.jpg','tcar3.jpg','tcar4.jpg']
```
[the remark this answers is not in the document]
Comment (ongzizhao) [reply tests/conftest.py:27]: done

File: peekingduck/pipeline/nodes/draw/blur_bbox.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        Function that blur the area bounded by bbox in an image
+        """
+        height = image.shape[0]
+        width = image.shape[1]
+
+        for bbox in bboxes:
+            x_1, y_1, x_2, y_2 = bbox
+            x_1, x_2 = int(x_1 * width), int(x_2 * width)
+            y_1, y_2 = int(y_1 * height), int(y_2 * height)
+
```
[the remark this answers is not in the document]
Comment (chngzyk) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py:58]: Can you point me to the scaling function?

[the remark this answers is not in the document]
Comment (chngzyk) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py:58]: Btw, this segment of the code is also causing lint error. I think it is caused it has the same code as the YOLOV4-face. Do consider adding pylint ignore in this segment.

[the remark this answers is not in the document]
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py:58]: In the license_plate_files.detector.py there is a function called bbox_scaling.

l added the pylint ignore comment for duplicate code

File: peekingduck/pipeline/nodes/draw/blur_bbox.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+        |img|
+
+    Configs:
+        None.
+    """
+
+    def __init__(self, config: Dict[str, Any] = None, **kwargs: Any) -> None:
+        super().__init__(config, node_path=__name__, **kwargs)
+
+    # pylint: disable=duplicate-code
```
Comment (chngzyk) [inline peekingduck/pipeline/nodes/draw/blur_bbox.py:46]: Did this line of code manage to suppress the warning? If no, remove this line.

File: peekingduck/pipeline/nodes/draw/blur_bbox.py [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+Blur area bounded by bounding boxes over detected object
+"""
+
+from typing import Any, Dict, List
+import numpy as np
+from scipy.ndimage import gaussian_filter
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+
+class Node(AbstractNode):  # pylint: disable=too-few-public-methods, duplicate-code
```
Comment (chngzyk) [inline peekingduck/pipeline/nodes/draw/blur_bbox.py:24]: Does the "duplicate-code" suppress the warning? If no, remove

Comment (ongzizhao) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py]: the comment is removed, as it does not work

File: peekingduck/pipeline/nodes/draw/blur_bbox.py (hunk 16 above)
Comment (ongzizhao) [reply peekingduck/pipeline/nodes/draw/blur_bbox.py]: the pylint: disable=duplicate-code comment is removed, as it does not work

**Pull request merged**